File: OPass/Tabs/ScheduleTab/ScheduleTableViewController.m
//
//  ScheduleTableViewController.m
//  CCIP
//
//  Created by FrankWu on 2017/7/16.
//  Copyright © 2017年 CPRTeam. All rights reserved.
//

#import "ScheduleTableViewController.h"
#import "ScheduleTableViewCell.h"
#import "ScheduleDetailViewController.h"
#import "AppDelegate.h"

@interface ScheduleTableViewController ()

@property (strong, nonatomic) NSMutableArray *programTimes;
@property (strong, nonatomic) NSMutableDictionary *programSections;

@end

@implementation ScheduleTableViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self registerForceTouch];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;
    self.programTimes = [NSMutableArray new];
    self.programSections = [NSMutableDictionary new];
    for (NSDictionary *program in self.programs) {
        NSDate *startTime = [Constants DateFromString:[program objectForKey:@"start"]];
        NSString *start = [Constants DateToDisplayTimeString:startTime];
        NSMutableArray *section = [self.programSections objectForKey:start];
        if (section == nil) {
            section = [NSMutableArray new];
            [self.programTimes addObject:startTime];
        }
        [section addObject:program];
        [self.programSections setObject:section
                                 forKey:start];
    }
    NSSortDescriptor *sortDescriptor = [[NSSortDescriptor alloc] initWithKey:@"timeIntervalSince1970"
                                                                   ascending:YES];
    [self.programTimes sortUsingDescriptors:[NSArray arrayWithObject:sortDescriptor]];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    [self.tableView reloadData];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (NSArray<id<UIPreviewActionItem>> *)previewActionItems {
    return [self previewActions];
}

- (UIViewController *)previewingContext:(id<UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location {
    UITableView *tableView = (UITableView *)[previewingContext sourceView];
    NSIndexPath *indexPath = [((NSArray *)[tableView valueForKey:@"_highlightedIndexPaths"]) firstObject];
    if (indexPath != nil) {
        UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Schedule"
                                                             bundle:nil];
        ScheduleDetailViewController *detailView = [storyboard instantiateViewControllerWithIdentifier:INIT_SCHEDULE_DETAIL_VIEW_STORYBOARD_ID];
        NSDate *time = [self.programTimes objectAtIndex:indexPath.section];
        NSString *timeString = [Constants DateToDisplayTimeString:time];
        NSDictionary *program = [[self.programSections objectForKey:timeString] objectAtIndex:indexPath.row];
        [detailView setDetailData:program];
        UITableViewCell *tableCell = [tableView cellForRowAtIndexPath:indexPath];
        [previewingContext setSourceRect:[self.view convertRect:tableCell.frame fromView:tableView]];
        return detailView;
    } else {
        return nil;
    }
}

- (void)previewingContext:(id<UIViewControllerPreviewing>)previewingContext commitViewController:(UIViewController *)viewControllerToCommit {
    [self.navigationController showViewController:viewControllerToCommit sender:nil];
}

#pragma mark - Table view data source
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section {
    return 40.0f;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 80.0f;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [self.programSections count];
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return [Constants DateToDisplayTimeString:[self.programTimes objectAtIndex:section]];
}

- (void)tableView:(UITableView *)tableView willDisplayHeaderView:(UIView *)view forSection:(NSInteger)section {
    UITableViewHeaderFooterView *header = (UITableViewHeaderFooterView *)view;
    [header.textLabel setTextColor:[AppDelegate AppConfigColor:@"ScheduleSectionTitleTextColor"]];
    [view setTintColor:[AppDelegate AppConfigColor:@"ScheduleSectionTitleBackgroundColor"]];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    NSDate *time = [self.programTimes objectAtIndex:section];
    NSString *timeString = [Constants DateToDisplayTimeString:time];
    return [[self.programSections objectForKey:timeString] count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *scheduleCellName = @"ScheduleCell";
    
    ScheduleTableViewCell *cell = (ScheduleTableViewCell *)[tableView dequeueReusableCellWithIdentifier:scheduleCellName];
    if (cell == nil) {
        [tableView registerNib:[UINib nibWithNibName:@"ScheduleTableViewCell" bundle:nil] forCellReuseIdentifier:scheduleCellName];
        cell = (ScheduleTableViewCell *)[tableView dequeueReusableCellWithIdentifier:scheduleCellName];
    }
    
    NSDate *time = [self.programTimes objectAtIndex:indexPath.section];
    NSString *timeString = [Constants DateToDisplayTimeString:time];
    NSDictionary *program = [[self.programSections objectForKey:timeString] objectAtIndex:indexPath.row];
    [cell setDelegate:self];
    [cell setSchedule:program];
    [cell setSelectionStyle:UITableViewCellSelectionStyleGray];
    
    NSDate *endTime = [Constants DateFromString:[program objectForKey:@"end"]];
    NSTimeInterval sinceEnd = [endTime timeIntervalSinceDate:self.pagerController.today];
    [cell setDisabled:(sinceEnd < 0)];
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath
                             animated:YES];
    NSDate *time = [self.programTimes objectAtIndex:indexPath.section];
    NSString *timeString = [Constants DateToDisplayTimeString:time];
    NSDictionary *program = [[self.programSections objectForKey:timeString] objectAtIndex:indexPath.row];
    [self.pagerController performSegueWithIdentifier:SCHEDULE_DETAIL_VIEW_STORYBOARD_ID
                                              sender:program];
}

- (NSString *)getID:(NSDictionary *)program {
    return [NSString stringWithFormat:@"%@-%@-%@", [program objectForKey:@"room"], [program objectForKey:@"start"], [program objectForKey:@"end"]];
}

- (void)actionFavorite:(NSString *)scheduleId {
    NSDictionary *favProgram = @{};
    NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
    NSObject *favObj = [userDefault valueForKey:FAV_KEY];
    NSArray *favoriteArray = [favObj isKindOfClass:[NSData class]] ? [NSKeyedUnarchiver unarchiveObjectWithData:(NSData *)favObj] : favObj;
    NSMutableArray *favorites = [NSMutableArray arrayWithArray:favoriteArray];
    for (NSDate *time in self.programTimes) {
        NSString *timeString = [Constants DateToDisplayTimeString:time];
        for (NSDictionary *program in [self.programSections objectForKey:timeString]) {
            if (program != nil && [[self getID:program] isEqualToString:scheduleId]) {
                favProgram = program;
                break;
            }
        }
        if ([[favProgram allKeys] count] > 0) {
            break;
        }
    }
    BOOL hasFavorite = [self hasFavorite:scheduleId];
    if (!hasFavorite) {
        [favorites addObject:favProgram];
    } else {
        [favorites removeObject:favProgram];
    }
    NSData *favData = [NSKeyedArchiver archivedDataWithRootObject:favorites];
    [userDefault setValue:favData
                   forKey:FAV_KEY];
    [userDefault synchronize];
    [self.tableView reloadData];
//    [OPassAPI RegisteringFavoriteScheduleForEvent:[Constants EventId]
//                                        withToken:[Constants AccessToken]
//                                       toSchedule:scheduleId
//                                        isDisable:NO
//                                       completion:^(BOOL success, id _Nullable obj, NSError * _Nonnull error) {
//                                           NSLog(@"%@", obj);
//                                       }];
}

- (BOOL)hasFavorite:(NSString *)scheduleId {
    NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
    // ugly convension for crash prevent
    NSObject *favObj = [userDefault valueForKey:FAV_KEY];
    NSArray *favorites = [favObj isKindOfClass:[NSData class]] ? [NSKeyedUnarchiver unarchiveObjectWithData:(NSData *)favObj] : favObj;
    for (NSDictionary *program in favorites) {
        if ([[self getID:program] isEqualToString:scheduleId]) {
            return YES;
        }
    }
    return NO;
}

@end

File: OPass/Tabs/ScheduleTab/ScheduleTableViewController.swift
//
//  ScheduleTableViewController.swift
//  OPass
//
//  Created by 腹黒い茶 on 2019/6/13.
//  Copyright © 2019 OPass. All rights reserved.
//

import Foundation
import UIKit

class ScheduleTableViewController: UITableViewController, UIViewControllerPreviewingDelegate, ScheduleFavoriteDelegate {
    func getID(_ program: NSDictionary) -> String {
        return ""
    }

    func actionFavorite(_ scheduleId: String) {

    }

    func hasFavorite(_ scheduleId: String) -> Bool {
        return false
    }

    //- (void)actionFavorite:(NSString *)scheduleId {
    //    NSDictionary *favProgram = @{};
    //    NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
    //    NSObject *favObj = [userDefault valueForKey:Constants.FAV_KEY];
    //    NSArray *favoriteArray = [favObj isKindOfClass:[NSData class]] ? [NSKeyedUnarchiver unarchiveObjectWithData:(NSData *)favObj] : favObj;
    //    NSMutableArray *favorites = [NSMutableArray arrayWithArray:favoriteArray];
    //    for (NSDate *time in self.programTimes) {
    //        NSString *timeString = [Constants DateToDisplayTimeString:time];
    //        for (NSDictionary *program in [self.programSections objectForKey:timeString]) {
    //            if (program != nil && [[self getID:program] isEqualToString:scheduleId]) {
    //                favProgram = program;
    //                break;
    //            }
    //        }
    //        if ([[favProgram allKeys] count] > 0) {
    //            break;
    //        }
    //    }
    //    BOOL hasFavorite = [self hasFavorite:scheduleId];
    //    if (!hasFavorite) {
    //        [favorites addObject:favProgram];
    //    } else {
    //        [favorites removeObject:favProgram];
    //    }
    //    NSData *favData = [NSKeyedArchiver archivedDataWithRootObject:favorites];
    //    [userDefault setValue:favData
    //                   forKey:Constants.FAV_KEY];
    //    [userDefault synchronize];
    //    [self.tableView reloadData];
    ////    [OPassAPI RegisteringFavoriteScheduleForEvent:[Constants EventId]
    ////                                        withToken:[Constants AccessToken]
    ////                                       toSchedule:scheduleId
    ////                                        isDisable:NO
    ////                                       completion:^(BOOL success, id _Nullable obj, NSError * _Nonnull error) {
    ////                                           NSLog(@"%@", obj);
    ////                                       }];
    //}
    //
    //- (BOOL)hasFavorite:(NSString *)scheduleId {
    //    NSUserDefaults *userDefault = [NSUserDefaults standardUserDefaults];
    //    // ugly convension for crash prevent
    //    NSObject *favObj = [userDefault valueForKey:Constants.FAV_KEY];
    //    NSArray *favorites = [favObj isKindOfClass:[NSData class]] ? [NSKeyedUnarchiver unarchiveObjectWithData:(NSData *)favObj] : favObj;
    //    for (NSDictionary *program in favorites) {
    //        if ([[self getID:program] isEqualToString:scheduleId]) {
    //            return YES;
    //        }
    //    }
    //    return NO;
    //}

    public var pagerController: ScheduleViewPagerController?
    public var sessionIds: Array<String>?
    var programTimes = Array<Date>()
    var programSections = Dictionary<String, Array<String>>()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.registerForceTouch()
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        self.programTimes.removeAll()
        self.programSections.removeAll()
        for session in (self.pagerController?.programs!.Sessions.filter { (self.sessionIds?.contains($0.Id))! })! {
            let startTime = Constants.DateFromString(session.Start)
            let start = Constants.DateToDisplayTimeString(startTime)
            if self.programSections.index(forKey: start) == nil {
                self.programTimes.append(startTime)
                self.programSections[start] = Array<String>()
            }
            self.programSections[start]?.append(session.Id)
        }
        self.programTimes.sort()
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        self.tableView.reloadData()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    override var previewActionItems: [UIPreviewActionItem] {
        return self.previewActions()
    }

    func previewingContext(_ previewingContext: UIViewControllerPreviewing, viewControllerForLocation location: CGPoint) -> UIViewController? {
        let tableView = previewingContext.sourceView as! UITableView
        guard let indexPath = (tableView.value(forKey: "_highlightedIndexPaths") as! Array<IndexPath>).first else {
            return nil
        }
        let storyboard = UIStoryboard.init(name: "Schedule", bundle: nil)
        let detailView = storyboard.instantiateViewController(withIdentifier: Constants.INIT_SCHEDULE_DETAIL_VIEW_STORYBOARD_ID) as! ScheduleDetailViewController
        let time = Constants.DateToDisplayTimeString(self.programTimes[indexPath.section])
        let sessionId = (self.programSections[time]?[indexPath.row])!
        guard let session = self.pagerController?.programs!.GetSession(sessionId) else { return detailView }
        detailView.setSessionData(session)
        let tableCell = tableView.cellForRow(at: indexPath)
        previewingContext.sourceRect = self.view.convert(tableCell!.frame, from: tableView)
        return detailView
    }

    func previewingContext(_ previewingContext: UIViewControllerPreviewing, commit viewControllerToCommit: UIViewController) {
        self.navigationController?.show(viewControllerToCommit, sender: nil)
    }

    // MARK: - Table view data source

    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 40
    }

    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 80
    }

    override func numberOfSections(in tableView: UITableView) -> Int {
        return self.programSections.count
    }

    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return Constants.DateToDisplayTimeString(self.programTimes[section])
    }

    override func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
        let header = view as! UITableViewHeaderFooterView
        header.textLabel?.textColor = AppDelegate.appConfigColor("ScheduleSectionTitleTextColor")
        view.tintColor = AppDelegate.appConfigColor("ScheduleSectionTitleBackgroundColor")
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let time = Constants.DateToDisplayTimeString(self.programTimes[section])
        return self.programSections[time]!.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let scheduleCellName = "ScheduleCell"
        var cell = tableView.dequeueReusableCell(withIdentifier: scheduleCellName) as? ScheduleTableViewCell
        if cell == nil {
            tableView.register(UINib.init(nibName: "ScheduleTableViewCell", bundle: nil), forCellReuseIdentifier: scheduleCellName)
            cell = tableView.dequeueReusableCell(withIdentifier: scheduleCellName) as? ScheduleTableViewCell
        }

        let time = Constants.DateToDisplayTimeString(self.programTimes[indexPath.section])
        let sessionId = (self.programSections[time]?[indexPath.row])!
        guard let session = self.pagerController?.programs!.GetSession(sessionId) else { return cell! }
        let endTime = Constants.DateFromString(session.End)
        let sinceEnd = endTime.timeIntervalSince(self.pagerController!.today)

        cell?.selectionStyle = .gray
        cell?.setDisabled(sinceEnd < 0)
        cell?.setDelegate(self)
        cell?.setSession(session)

        return cell!
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let time = Constants.DateToDisplayTimeString(self.programTimes[indexPath.section])
        let sessionId = self.programSections[time]?[indexPath.row]
        self.pagerController?.performSegue(withIdentifier: Constants.SCHEDULE_DETAIL_VIEW_STORYBOARD_ID, sender: sessionId)
    }


}
